packages/w3p/src/providers/wrapped/evm-wallet-connect.ts

```diff
@@ -0,0 +1,316 @@
+{
+  #rawProvider: UniversalProvider
+  #ethProvider: providers.Web3Provider
+
+  #w3Modal: WalletConnectModal
+
+  #chainId?: ChainId
+  #address?: string
+
+  readonly #projectId: string
+
+  constructor(provider: RawProvider) {
+    super()
+
+    if (!(provider as WalletConnectInitArgs).projectId) {
+      throw new Error('projectId is required for WalletConnect provider')
+    }
+
+    const { projectId } = provider as WalletConnectInitArgs
+    this.#projectId = projectId
+    this.#rawProvider = {} as UniversalProvider
+    this.#ethProvider = {} as providers.Web3Provider
+    this.#w3Modal = new WalletConnectModal({
+      projectId: this.#projectId,
+    })
+  }
+
+  static get providerType(): PROVIDERS {
+    return PROVIDERS.WalletConnect
+  }
+
+  get rawProvider() {
+    return this.#rawProvider as unknown as RawProvider
+  }
+
+  get chainType(): CHAIN_TYPES {
+    return CHAIN_TYPES.EVM
+  }
+
+  get isConnected(): boolean {
+    return Boolean(this.#chainId) || Boolean(this.#address)
+  }
+
+  get chainId(): ChainId | undefined {
+    return this.#chainId
+  }
+
+  get address(): string | undefined {
+    return this.#address
+  }
+
+  get #defaultEventPayload() {
+    return {
+      address: this.#address,
+      chainId: this.#chainId,
+      isConnected: this.isConnected,
+    }
+  }
+
+  async init(): Promise<void> {
+    this.#rawProvider = await UniversalProvider.init({
+      logger: 'info',
+      relayUrl: 'wss://relay.walletconnect.com',
+      projectId: this.#projectId,
+    })
+
+    await this.#checkForPersistedSession()
+
+    await this.#setListeners()
+
+    this.emit(PROVIDER_EVENT_BUS_EVENTS.Initiated, this.#defaultEventPayload)
+  }
+
+  async connect(): Promise<void> {
+    await this.#rawProvider.connect({
+      namespaces: {
+        eip155: {
+          methods: [
+            'wallet_switchEthereumChain',
+            'wallet_addEthereumChain',
+            'eth_requestAccounts',
+
+            'eth_sendTransaction',
+
+            'eth_sign',
+            'personal_sign',
+            'eth_signTypedData',
+          ],
+          chains: ['eip155:1'],
+          events: [
+            PROVIDER_EVENTS.Connect,
+            PROVIDER_EVENTS.Disconnect,
+            PROVIDER_EVENTS.ChainChanged,
+            PROVIDER_EVENTS.AccountsChanged,
+          ],
+        },
+      },
+      optionalNamespaces: {
+        eip155: {
+          methods: [
+            'wallet_switchEthereumChain',
+            'wallet_addEthereumChain',
+            'eth_requestAccounts',
+
+            'eth_sendTransaction',
+
+            'eth_sign',
+            'personal_sign',
+            'eth_signTypedData',
+          ],
+          chains: ['eip155:1'],
+          events: [
+            PROVIDER_EVENTS.Connect,
+            PROVIDER_EVENTS.Disconnect,
+            PROVIDER_EVENTS.ChainChanged,
+            PROVIDER_EVENTS.AccountsChanged,
+          ],
+        },
+      },
+      skipPairing: true,
```

Comment:
move creating this object to the helpers

Reply:
done